std/experimental/checkedint.d

```diff
@@ -0,0 +1,1645 @@
+        else
+            return lhs;
+    }
+
+    typeof(lhs ^^ rhs) b = void;
+    static if (!isUnsigned!L && isUnsigned!(typeof(b)))
+    {
+        // Need to worry about mixed-sign stuff
+        if (lhs < 0)
+        {
+            if (rhs & 1)
+            {
+                if (lhs < 0) overflow = true;
+                return 0;
+            }
+            b = -lhs;
+        }
+        else
+        {
+            b = lhs;
+        }
+    }
+    else
+    {
+        b = lhs;
+    }
+    if (b == 1) return 1;
+    if (b == -1) return (rhs & 1) ? -1 : 1;
+    if (rhs > 63)
+    {
+        overflow = true;
+        return 0;
+    }
+
+    assert((b > 1 || b < -1) && rhs > 1);
+    return powImpl(b, cast(uint) rhs, overflow);
+}
+
+// Inspiration: http://www.stepanovpapers.com/PAM.pdf
+pure @safe nothrow @nogc
+private T powImpl(T)(T b, uint e, ref bool overflow)
+if (isIntegral!T && T.sizeof >= 4)
+{
+    assert(e > 1);
+
+    import core.checkedint : muls, mulu;
+    static if (isUnsigned!T) alias mul = mulu;
+    else alias mul = muls;
+
+    T r = b;
+    --e;
+    // Loop invariant: r * (b ^^ e) is the actual result
+    for (;; e /= 2)
+    {
+        if (e % 2)
+        {
+            r = mul(r, b, overflow);
+            if (e == 1) break;
+        }
+        b = mul(b, b, overflow);
+    }
+    return r;
+}
+
+unittest
+{
+    static void testPow(T)(T x, uint e)
+    {
+        bool overflow;
+        assert(opChecked!"^^"(T(0), 0, overflow) == 1);
+        assert(opChecked!"^^"(-2, T(0), overflow) == 1);
+        assert(opChecked!"^^"(-2, T(1), overflow) == -2);
+        assert(opChecked!"^^"(-1, -1, overflow) == -1);
+        assert(opChecked!"^^"(-2, 1, overflow) == -2);
+        assert(opChecked!"^^"(-2, -1, overflow) == 0);
+        assert(opChecked!"^^"(-2, 4u, overflow) == 16);
+        assert(!overflow);
+        assert(opChecked!"^^"(-2, 3u, overflow) == 0);
+        assert(overflow);
+        overflow = false;
+        assert(opChecked!"^^"(3, 64u, overflow) == 0);
+        assert(overflow);
+        overflow = false;
+        foreach (uint i; 0 .. e)
+        {
+            assert(opChecked!"^^"(x, i, overflow) == x ^^ i);
+            assert(!overflow);
+        }
+        assert(opChecked!"^^"(x, e, overflow) == x ^^ e);
+        assert(overflow);
+    }
+
+    testPow!int(3, 21);
+    testPow!uint(3, 21);
+    testPow!long(3, 40);
+    testPow!ulong(3, 41);
+}
+
+version(unittest) private struct CountOverflows
+{
+    uint calls;
+    auto onOverflow(string op, Lhs)(Lhs lhs)
+    {
+        ++calls;
+        return mixin(op~"lhs");
+    }
+    auto onOverflow(string op, Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        ++calls;
+        return mixin("lhs"~op~"rhs");
+    }
+    Lhs onBadOpOpAssign(string op, Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        ++calls;
+        return mixin("lhs"~op~"=rhs");
+    }
+}
+
+version(unittest) private struct CountOpBinary
```

Comment:
Is it really necessary not to define it directly in a unittest ? I see only one usage so far.